assets/js/googlesitekit/notifications/datastore/notifications.js

```diff
@@ -138,11 +192,78 @@ export const actions = {
 					isDismissible,
 					dismissRetries,
 					featureFlag,
-				},
-			},
-			type: REGISTER_NOTIFICATION,
-		};
-	},
+				};
+
+				// Check if the notification should be added to the queue
+				// before inserting it.
+				if (
+					! shouldNotificationBeAddedToQueue(
+						notification,
+						groupID,
+						undefined,
+						isNotificationDismissed
+					)
+				) {
+					return;
+				}
+
+				// To do this, we'll add the notification to the queue immediately,
+				// which will insert it in the right position based on its priority.
+				yield actions.insertNotificationIntoResolvedQueue( id );
+
+				return;
+			}
+
+			// If view contexts were provided, we need to repopulate the queues
+			// because the data store doesn’t know the current `viewContext` and thus
+			// can’t figure out if the ones specified are valid/active.
+			//
+			// This has the unfortunate side effect of causing rotation of the
+			// notifications in the queue, so specifying `viewContexts` should
+			// be avoided when registering "ad-hoc" notifications that aren't
+			// registered on initialization.
+			//
+			// For each view context, check to see if the notifications have
+			// finished resolution for the `getQueuedNotifications` selector.
+			//
+			// If they have, we need to invalidate the `getQueuedNotifications`
+			// resolver for that `viewContext` + `groupID` combination, as it's
+			// no longer valid.
+			//
+			// Note that this is an unlikely scenario, as notifications that are
+			// registered after initial page load are usually ad-hoc notifications
+			// that do not specify a `viewContext` (because they're intended to be
+			// immediately visible).
+			//
+			// Still: this code is here to ensure that if a notification _is_
+			// registered after initial page load, it will be visible in the queue
+			// if appropriate.
+			viewContexts.forEach( ( viewContext ) => {
+				const hasResolvedGetQueuedNotifications = registry
+					.select( CORE_NOTIFICATIONS )
+					.hasFinishedResolution( 'getQueuedNotifications', [
```

Comment:
We could destructure `hasFinishedResolution` once before iterating to avoid multiple registry selects.